### PR TITLE
add controller flag options to KubeVirt CR

### LINKED ISSUE #2746: Make controller threads configurable on the operator CR
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug

/kind enhancement

**What happened**:

Some users fine-tune the controller thread count. See #2648. Using manifests is already deprecated. We should not forget to move these options  to the operator install.

**What you expected to happen**:

That I can either set these specific options on the CR, or that we introduce a generic mapping between component commandline arguments and the CR, so that we don't have to map each one individually.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`):
- Kubernetes version (use `kubectl version`):
- VM or VMI specifications:
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by ashleyschuett**
**What this PR does / why we need it**:

This PR allows users to define flags values for running the KubeVirt Controller in the KubeVirt CR. 
This is needed so that users that are deploying KubeVirt using the operator can set flags on the controller at creation time. 

**Which issue(s) this PR fixes**: 
Fixes #2746 

**Special notes for your reviewer**:

**Release note**:
```release-note
Set KubeVirt controller flags in the KubeVirt CR by specifying the key ControllerFlags key with a map of key values. 
```

### CODE REVIEW

File: pkg/virt-operator/util/config.go [hunk 1]
```diff
@@ -419,6 +426,22 @@ func (c *KubeVirtDeploymentConfig) GetVerbosity() string {
...
+	return c.ControllerFlags
+}
+
+func (c *KubeVirtDeploymentConfig) GetControllerFlagsAsArray() []string {
+	flags := make([]string, 0)
+
+	for f, v := range c.GetControllerFlags() {
+		str := matchAllCap.ReplaceAllString(f, "${1}-${2}")
+		flags = append(flags, fmt.Sprintf("--%s", strings.ToLower(str)))
+		flags = append(flags, v)
```
Comment (davidvossel) [inline pkg/virt-operator/util/config.go:439]: what happens if a boolean flag is added? Then there wouldn't be a value to set it to true, right?

maybe ignore appending the value to the slice `if v == ""` ?

File: pkg/virt-operator/util/config.go [hunk 2]
```diff
@@ -419,6 +426,22 @@ func (c *KubeVirtDeploymentConfig) GetVerbosity() string {
...
 
+func (c *KubeVirtDeploymentConfig) GetControllerFlags() map[string]string {
+	return c.ControllerFlags
+}
+
+func (c *KubeVirtDeploymentConfig) GetControllerFlagsAsArray() []string {
+	flags := make([]string, 0)
+
+	for f, v := range c.GetControllerFlags() {
+		str := matchAllCap.ReplaceAllString(f, "${1}-${2}")
```
Comment (davidvossel) [inline pkg/virt-operator/util/config.go:437]: why do we need this  regex and to convert convert to lower case?

File: pkg/virt-operator/util/config.go [hunk 3]
```diff
@@ -99,6 +101,9 @@ type KubeVirtDeploymentConfig struct {
 
 	// environment variables from virt-operator to pass along
 	PassthroughEnvVars map[string]string `json:"passthroughEnvVars,omitempty" optional:"true"`
+
+	// generic mapping for the controller deployments flags
+	ControllerFlags map[string]string `json:"virtControllerFlags,omitempty" optional:"true"`
```
Comment (davidvossel) [inline pkg/virt-operator/util/config.go:106]: can we call this VirtControllerFlags to align with the json field. That's a convention we generally try to adhere to.

File: staging/src/kubevirt.io/client-go/api/v1/types.go [hunk 4]
```diff
@@ -1084,6 +1084,9 @@ type KubeVirtCertificateRotateStrategy struct {
 // ---
 // +k8s:openapi-gen=true
 type KubeVirtSpec struct {
+	// Configure the value used for each flag in the KubeVirt controller.
+	ControllerFlags map[string]string `json:"controllerFlags,omitempty"`
```
Comment (davidvossel) [inline staging/src/kubevirt.io/client-go/api/v1/types.go:1088]: maybe call this VirtControllerFlags for consistency?

File: pkg/virt-operator/util/config.go (hunk 2 above)
Comment (ashleyschuett) [reply pkg/virt-operator/util/config.go:437]: My original thought here was `userControllerThreads` -> `user-controller-threads`. 

But really the map should look like such
```
VirtControllerFlags: 
  user-controller-threads: "3"
```

I'll remove this.

**Comment by davidvossel** [comment]
I'd like to see a conclusion to this issue https://github.com/kubevirt/kubevirt/issues/3365 before we commit to an api change.

The approach here is technically fine. We're getting requests to do more detailed modifications outside of just controller args though, which might require a more generic solution

**Comment by rmohr** [comment]
In the context of the discussion regarding to allowing json patches on our resources, I guess we want to wait with this PR, right?

**Comment by ashleyschuett** [comment]
Yes, I believe we want to wait. The only thing that was mentioned is that if this is potentially better user experience for something that would be configured often we might want to support both. I can close this PR for now though, if we want to wait on a conclusion from the discussion?

**Comment by rmohr** [comment]
/hold

> Yes, I believe we want to wait. The only thing that was mentioned is that if this is potentially better user experience for something that would be configured often we might want to support both. I can close this PR for now though, if we want to wait on a conclusion from the discussion?

Letting it open works for me. Let's just put a hold there until we know the way forward from the discussion in https://github.com/kubevirt/kubevirt/issues/3365.

**Comment by ashleyschuett** [comment]
based on conversation with @rmohr it sounds like we want to continue to pursue configuring deployments through patches and continue the discussion in the previous linked issue, but also have the option to easily configure the flags on the controller as this PR does. If this is correct can we remove the `hold` label and rereview?

**Comment by rmohr** [comment]
> based on conversation with @rmohr it sounds like we want to continue to pursue configuring deployments through patches and continue the discussion in the previous linked issue, but also have the option to easily configure the flags on the controller as this PR does. If this is correct can we remove the `hold` label and rereview?

My line of thought was that it may make sense to make them not part of the "installation specification". I though about moving them to the `kubevirt-config`. They would then be configured similar to other values for components. The component (virt-controller) would then simply pick it up. The PR would then be more connected to #3455 (config propagation) and less to defining an install-strategy which requires modifying specs and rolling them out.

@davidvossel you ok with this?

**Comment by davidvossel** [comment]
> My line of thought was that it may make sense to make them not part of the "installation specification". I though about moving them to the kubevirt-config. They would then be configured similar to other values for components. The component (virt-controller) would then simply pick it up. The PR would then be more connected to #3455 (config propagation) and less to defining an install-strategy which requires modifying specs and rolling them out.

we agree the config propagation needs to take place and we agree the logic around supporting patches needs to take place as well.

IMO, I'd move forward with the config propagation and patch features first as they solve an entire class of customization issues we have today, including the one in this PR. I see iterating after that to expose new options on the kubevirt CR focused on virt-controller customizations as secondary. 

So, I'm not opposed to what you're saying at all, I guess I'm just wanting to clarify the priority here of what we're doing first.

**Comment by rmohr** [comment]
@ashleyschuett do you think we can revive this PR? It would be great if we directly pass through the flags to the components and they ideally adjust themselves (maybe even without restart).

**Comment by ashleyschuett** [comment]
> @ashleyschuett do you think we can revive this PR? It would be great if we directly pass through the flags to the components and they ideally adjust themselves (maybe even without restart).

@rmohr I updated this PR before and had forgotten about this comment. Would that mean the flag values would be adjusted in the actual component and not updated in the deployment spec? The new PR is https://github.com/kubevirt/kubevirt/pull/5563 should I close this PR and we continue the conversation there?

**Pull request closed without merging**